DIPs/DIP1006.md

```diff
@@ -0,0 +1,101 @@
+| RC#             | 0                                                               |
+| Author:         | Mathias Lang - mathias.lang@sociomantic.com                     |
+| Implementation: | None                                                            |
+| Status:         | Will be set by the DIP manager (e.g. "Approved" or "Rejected")  |
+
+
+## Abstract
+
+A proposal to expose a way to selectively disable some of D's contract features
+(invariant, `in` / `out` and `assert`).
+
+
+### Links
+
+Limited D1 implementation, used by Sociomantic: https://github.com/dlang/dmd/pull/6347
+
+
+## Description
+
+Add a command-line switch, named `-contracts` in this proposal, which takes 4 possible values:
+
+- `all`: the default, present for completeness and scripts, so one can just append it to a generated command line to enable all contracts checks.
+- `noinvariant`: Disable implicit `invariant` calls, but leaves function's `in` / `out` contracts and `assert` in.
+- `assert`: Disable `invariant` blocks and `in` / `out` contracts, but leaves `assert` in.
+- `none`: Disable all `invariant`, `in` and `out` contracts, and `assert`.
+
+Note that the `noinvariant` value only disables implicit calls to invariant. Explicit calls can only disabled by disabling `assert`.
+
+
+### Rationale
+
+#### Prelude on contracts
+
+Under the broad name "Contracts", we regroup 4 of D features: `assert`, `in` contracts, `out` contracts and `invariant`.
+Contracts in D are a mean to do sanity checking. A contract that fails is a sign of a faulty logic in the code.
+Since contracts should never fail, the currently advertised approach for D programers is to use them during development,
+then deploy a `-release` build of the application for maximum performances.
+
+Contracts naturally fit in a hierarchy, where `assert` is the most basic block (being used by others),
+and `invariant` is meant for expensive sanity checking. `in` and `out` contracts are located in between.
+
+#### Cost of contracts
+
+`assert` and `in` / `out` contracts follow a pay-for-what-you-use approach: an `assert` which is never hit is free,
+and a `in` or `out` contract on a function that is never called is also free of runtime overhead.
+
+`invariant` stand out in that regard: they can be used on a pay-for-what-you-use basis, by using `assert(object)` (or `assert(this)`),
+but they are also implicitly called twice on every `public` and `protected` class method call.
+The call itself is not direct: the compiler inserts a call to `_d_invariant`, [which can be found in druntime](https://github.com/dlang/druntime/blob/v2.072.0/src/rt/invariant.d).
+The code being very simple results in the class hierarchy to be traversed completely twice (no caching is done)
+for every call, even for classes which do not define any invariant.
+
+Experience showed that simply disabling this expensive lookup method can easily lead to 20% more throughput for OOP-intensive codebase.
```

Comment:
I would be interesting to also show some numbers for a public project, so it can be checked and peer-reviewed.